Memory allocation fails on Windows 7 SP1

#### What happens?

DuckDB fails when trying to allocate memory on Windows 7.

#### To Reproduce

Grab the DuckDB Windows executable. I tried both the 32- and the 64-bit CLI variants, and also tried the 64-bit variant from Python 3.8.

```console
$ duckdb.exe
v0.3.2 5aebf7dac
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D select 42;
Error: unable to open database ":memory:": IO Error: Could not fetch available system memory!
```

The same error occurs when trying to use a database file except `:memory:` is replaced by the database file's name in the error message.

I'm sure running on the Windows 7 OS is now quite rare among DuckDB users. So I'm mainly raising this issue so others can find it if they stumble upon it.

#### Environment (please complete the following information):
 - OS: Windows 7 SP1 running in VirtualBox on a Fedora 35 host
 - DuckDB Version: 0.3.2
 - DuckDB Client: CLI

#### Before Submitting

- [ ] **Have you tried this on the latest `master` branch?** no

I have to add that I'm running Windows 7 in VirtualBox on a Fedora 35 host. I edited the original post to add this information.

I tried building a fresh DuckDB binary with Cygwin, CMake and GCC 11. However, the build failed:

```console
Consolidate compiler generated dependencies of target duckdb_common
[ 33%] Building CXX object src/common/CMakeFiles/duckdb_common.dir/ub_duckdb_common.cpp.o
In file included from /cygdrive/c/Users/Gabor/Downloads/duckdb/build/src/common/ub_duckdb_common.cpp:16:
/cygdrive/c/Users/Gabor/Downloads/duckdb/src/common/local_file_system.cpp: In member function ‘virtual void duckdb::LocalFileSystem::Read(duckdb::FileHandle&, void*, int64_t, duckdb::idx_t)’:
/cygdrive/c/Users/Gabor/Downloads/duckdb/src/common/local_file_system.cpp:203:30: error: ‘pread’ was not declared in this scope; did you mean ‘read’?
  203 |         int64_t bytes_read = pread(fd, buffer, nr_bytes, location);
      |                              ^~~~~
      |                              read
/cygdrive/c/Users/Gabor/Downloads/duckdb/src/common/local_file_system.cpp: In member function ‘virtual void duckdb::LocalFileSystem::Write(duckdb::FileHandle&, void*, int64_t, duckdb::idx_t)’:
/cygdrive/c/Users/Gabor/Downloads/duckdb/src/common/local_file_system.cpp:224:33: error: ‘pwrite’ was not declared in this scope; did you mean ‘write’?
  224 |         int64_t bytes_written = pwrite(fd, buffer, nr_bytes, location);
      |                                 ^~~~~~
      |                                 write
/cygdrive/c/Users/Gabor/Downloads/duckdb/src/common/local_file_system.cpp: In member function ‘virtual void duckdb::LocalFileSystem::Truncate(duckdb::FileHandle&, int64_t)’:
/cygdrive/c/Users/Gabor/Downloads/duckdb/src/common/local_file_system.cpp:268:13: error: ‘ftruncate’ was not declared in this scope; did you mean ‘Truncate’?
  268 |         if (ftruncate(fd, new_size) != 0) {
      |             ^~~~~~~~~
      |             Truncate
make[2]: *** [src/common/CMakeFiles/duckdb_common.dir/build.make:76: src/common/CMakeFiles/duckdb_common.dir/ub_duckdb_common.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:4181: src/common/CMakeFiles/duckdb_common.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

--- a/src/common/file_system.cpp
+++ b/src/common/file_system.cpp
@@ -95,7 +95,7 @@ idx_t FileSystem::GetAvailableMemory() {
 	mem_state.dwLength = sizeof(MEMORYSTATUSEX);
 
 	if (GlobalMemoryStatusEx(&mem_state)) {
-		return MinValue<idx_t>(statex.ullTotalPhys, UINTPTR_MAX);
+		return MinValue<idx_t>(mem_state.ullTotalPhys, UINTPTR_MAX);
 	}
 	return DConstants::INVALID_INDEX;
 }

--- a/src/common/file_system.cpp
+++ b/src/common/file_system.cpp
@@ -59,7 +59,7 @@ idx_t FileSystem::GetAvailableMemory() {
 	errno = 0;
 	idx_t max_memory = MinValue<idx_t>((idx_t)sysconf(_SC_PHYS_PAGES) * (idx_t)sysconf(_SC_PAGESIZE), UINTPTR_MAX);
 	if (errno != 0) {
-		throw IOException("Could not fetch available system memory!");
+		return DConstants::INVALID_INDEX;
 	}
 	return max_memory;
 }
@@ -87,7 +87,7 @@ void FileSystem::SetWorkingDirectory(const string &path) {
 idx_t FileSystem::GetAvailableMemory() {
 	ULONGLONG available_memory_kb;
 	if (!GetPhysicallyInstalledSystemMemory(&available_memory_kb)) {
-		throw IOException("Could not fetch available system memory!");
+		return DConstants::INVALID_INDEX;
 	}
 	return MinValue<idx_t>(available_memory_kb * 1024, UINTPTR_MAX);
 }

--- a/src/common/file_system.cpp
+++ b/src/common/file_system.cpp
@@ -22,6 +22,7 @@
 #include <unistd.h>
 #else
 #include <string>
+#include <sysinfoapi.h>
 
 #ifdef __MINGW32__
 // need to manually define this for mingw
@@ -86,10 +87,17 @@ void FileSystem::SetWorkingDirectory(const string &path) {
 
 idx_t FileSystem::GetAvailableMemory() {
 	ULONGLONG available_memory_kb;
-	if (!GetPhysicallyInstalledSystemMemory(&available_memory_kb)) {
-		return DConstants::INVALID_INDEX;
+	if (GetPhysicallyInstalledSystemMemory(&available_memory_kb)) {
+		return MinValue<idx_t>(available_memory_kb * 1024, UINTPTR_MAX);
 	}
-	return MinValue<idx_t>(available_memory_kb * 1024, UINTPTR_MAX);
+	// fallback: try GlobalMemoryStatusEx
+	MEMORYSTATUSEX mem_state;
+	mem_state.dwLength = sizeof(MEMORYSTATUSEX);
+
+	if (GlobalMemoryStatusEx(&mem_state)) {
+		return MinValue<idx_t>(statex.ullTotalPhys, UINTPTR_MAX);
+	}
+	return DConstants::INVALID_INDEX;
 }
 
 string FileSystem::GetWorkingDirectory() {

--- a/src/include/duckdb/common/file_system.hpp
+++ b/src/include/duckdb/common/file_system.hpp
@@ -158,7 +158,7 @@ public:
 	DUCKDB_API static string GetWorkingDirectory();
 	//! Gets the users home directory
 	DUCKDB_API static string GetHomeDirectory();
-	//! Returns the system-available memory in bytes
+	//! Returns the system-available memory in bytes. Returns DConstants::INVALID_INDEX if the system function fails.
 	DUCKDB_API static idx_t GetAvailableMemory();
 	//! Path separator for the current file system
 	DUCKDB_API static string PathSeparator();

--- a/src/main/database.cpp
+++ b/src/main/database.cpp
@@ -205,7 +205,10 @@ void DatabaseInstance::Configure(DBConfig &new_config) {
 	}
 	config.maximum_memory = new_config.maximum_memory;
 	if (config.maximum_memory == (idx_t)-1) {
-		config.maximum_memory = FileSystem::GetAvailableMemory() * 8 / 10;
+		auto memory = FileSystem::GetAvailableMemory();
+		if (memory != DConstants::INVALID_INDEX) {
+			config.maximum_memory = memory * 8 / 10;
+		}
 	}
 	if (new_config.maximum_threads == (idx_t)-1) {
 #ifndef DUCKDB_NO_THREADS

